test/unit/sdam/server_selection.test.js

```diff
@@ -0,0 +1,381 @@
+      serverDescriptions.set('127.0.0.1:27018', secondary);
+
+      context('when the common server version is >= 5.0', function () {
+        const topologyDescription = new TopologyDescription(
+          TopologyType.ReplicaSetWithPrimary,
+          serverDescriptions,
+          'test',
+          MIN_SECONDARY_WRITE_WIRE_VERSION,
+          new ObjectId(),
+          MIN_SECONDARY_WRITE_WIRE_VERSION
+        );
+
+        context('when a read preference is provided', function () {
+          const selector = secondaryWritableServerSelector(
+            MIN_SECONDARY_WRITE_WIRE_VERSION,
+            ReadPreference.secondary
+          );
+          const server = selector(topologyDescription, Array.from(serverDescriptions.values()));
+
+          it('uses the provided read preference', function () {
+            expect(server).to.deep.equal([secondary]);
+          });
+        });
+
+        context('when a read preference is not provided', function () {
+          const selector = secondaryWritableServerSelector(MIN_SECONDARY_WRITE_WIRE_VERSION);
+          const server = selector(topologyDescription, Array.from(serverDescriptions.values()));
+
+          it('selects a primary', function () {
+            expect(server).to.deep.equal([primary]);
+          });
+        });
+      });
+
+      context('when the common server version is < 5.0', function () {
+        const topologyDescription = new TopologyDescription(
+          TopologyType.ReplicaSetWithPrimary,
+          serverDescriptions,
+          'test',
+          MIN_SECONDARY_WRITE_WIRE_VERSION - 1,
+          new ObjectId(),
+          MIN_SECONDARY_WRITE_WIRE_VERSION - 1
+        );
+
+        context('when a read preference is provided', function () {
+          const selector = secondaryWritableServerSelector(
+            MIN_SECONDARY_WRITE_WIRE_VERSION - 1,
+            ReadPreference.secondary
+          );
+          const server = selector(topologyDescription, Array.from(serverDescriptions.values()));
+
+          it('selects a primary', function () {
+            expect(server).to.deep.equal([primary]);
+          });
+        });
+
+        context('when read preference is not provided', function () {
+          const selector = secondaryWritableServerSelector(MIN_SECONDARY_WRITE_WIRE_VERSION - 1);
+          const server = selector(topologyDescription, Array.from(serverDescriptions.values()));
+
+          it('selects a primary', function () {
+            expect(server).to.deep.equal([primary]);
+          });
+        });
+      });
+
+      context('when a common wire version is not provided', function () {
+        const topologyDescription = new TopologyDescription(
+          TopologyType.ReplicaSetWithPrimary,
+          serverDescriptions,
+          'test',
+          MIN_SECONDARY_WRITE_WIRE_VERSION,
+          new ObjectId(),
+          MIN_SECONDARY_WRITE_WIRE_VERSION
+        );
+        const selector = secondaryWritableServerSelector();
```

Comment:
this does not differentiate between the wire version not being provided and the read preference not being provided, so I think doing `(undefined, ReadPreference.secondary)` would work better

Reply:
I've updated to this now. Nice suggestion - this actually caught a bug.